[ML] Make Elser model version configurable

davidkyle opened the pull request:
Adds a service option to make the version of the ELSER model used configurable. The options are `.elser_model_1`, `.elser_model_2`, `.elser_model_2_linux-x86_64` or the default if not specified. 

The logic for picking the default model is this:
1. If all the machine learning nodes are Linux x86 nodes then use the platform optimised version`.elser_model_2_linux-x86_64`
2. If running in Elastic cloud then use platform optimised version`.elser_model_2_linux-x86_64`
3. Otherwise use `.elser_model_2`

Labeled as a `non-issue` it is fixed a problem in unreleased code

davidkyle committed:
Make Elser model version configurable

davidkyle committed:
check style

davidkyle committed:
fix the tests

droberts195 commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elser/ElserMlNodeServiceSettings.java:
```diff
@@ -61,21 +70,36 @@ public static ElserMlNodeServiceSettings fromMap(Map<String, Object> map) {
+                validationException.addValidationError("unknown ELSER model version [" + version + "]");
+            }
+        } else {
+            version = ElserMlNodeService.ELSER_V2_MODEL;
```
I think this needs to choose a cleverer default. Otherwise this API will be out-of-sync with what we're doing in the UI with defaulting to the most efficient model.

The logic should be:

1. Use `ELSER_V2_MODEL_LINUX_X86` if a set containing the single item `linux-x86_64` is returned by https://github.com/elastic/elasticsearch/blob/ab93bfd972f3ba65046dc823b0f9bf174a43a1e7/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/MlPlatformArchitecturesUtil.java#L34
2. Otherwise, if that set came back empty and running in Cloud use `ELSER_V2_MODEL_LINUX_X86`
3. Otherwise, use `ELSER_V2_MODEL`

That will then match the UI logic.

So it creates a need to know if we're in Cloud. We'll have to use some existing cluster setting as a heuristic for detecting this. I think `xpack.ml.use_auto_machine_memory_percent` being `true` is the best we've got.

davidkyle replied:
Done. This caused a little churn as the `xpack.ml.use_auto_machine_memory_percent` setting and `MlPlatformArchitecturesUtil` had to be moved to core to be accessible.

davidkyle committed:
Pick the optimised model if possible

# Conflicts:
#	x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferencePlugin.java
#	x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elser/ElserMlNodeServiceSettings.java

davidkyle committed:
Merge branch 'main' into elser-2

davidkyle committed:
check style

droberts195 reviewed (approved):
LGTM - thanks for doing all those changes

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.11  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 100345`
